ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -298,18 +306,30 @@ void fillChunks() {
             channelReadBuffer = chunkFillerChannel.getNextChunk(0);
           }
           if (channelReadBuffer != null) {
+            maybeStopTrackingWaitForChannelDataTime();
             chunkToFill = getChunkToFill();
             if (chunkToFill == null) {
-              // no chunks are free to be filled yet.
+              // channel has data, but no chunks are free to be filled yet.
+              maybeStartTrackingWaitForChunkTime();
               break;
-            }
-            bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
-            if (!channelReadBuffer.hasRemaining()) {
-              chunkFillerChannel.resolveOldestChunk(null);
-              channelReadBuffer = null;
+            } else {
+              // channel has data, and there is a chunk that can be filled.
+              maybeStopTrackingWaitForChunkTime();
+              bytesFilledSoFar += chunkToFill.fillFrom(channelReadBuffer);
+              if (chunkToFill.isReady()) {
+                updateChunkFillerWaitTimeMetrics();
+              }
+              if (!channelReadBuffer.hasRemaining()) {
+                chunkFillerChannel.resolveOldestChunk(null);
+                channelReadBuffer = null;
+              }
             }
           } else {
             // channel does not have more data yet.
+            if (getFreeChunk() != null) {
+              // this means there is a chunk available to be filled, but no data in the channel.
+              maybeStartTrackingWaitForChannelDataTime();
```

Comment:
Brought this back, as although this won't be an accurate metric, it will still be useful and would account "mostly" for the wait times. If the data becomes available in the 100th iteration after a wait, then the time taken by the chunk filler thread outside of this operation before entering each of those 99 iterations do not affect this metric.